pkg/restore/restorers/pv_restorer.go

```diff
@@ -56,66 +57,46 @@ func (sr *persistentVolumeRestorer) Prepare(obj runtime.Unstructured, restore *a
+		// The backup had snapshots disabled, so we can return early
+		return obj, nil, nil
 	}
 
-	// if it's an unsupported volume type for snapshot restores, we're done
-	if sourceType, _ := kubeutil.GetPVSource(spec); sourceType == "" {
+	if boolptr.IsSetToFalse(restore.Spec.RestorePVs) {
+		// The restore has pv restores disabled, so we can return early
 		return obj, nil, nil
 	}
 
-	restoreFromSnapshot := false
-
-	if restore.Spec.RestorePVs != nil && *restore.Spec.RestorePVs {
-		// when RestorePVs = yes, it's an error if we don't have a snapshot service
-		if sr.snapshotService == nil {
-			return nil, nil, errors.New("PV restorer is not configured for PV snapshot restores")
-		}
+	// Past this point, we expect to be doing a restore
 
-		// if there are no snapshots in the backup, return without error
-		if backup.Status.VolumeBackups == nil {
-			return obj, nil, nil
-		}
-
-		// if there are snapshots, and this is a supported PV type, but there's no
-		// snapshot for this PV, it's an error
-		if backup.Status.VolumeBackups[pvName] == nil {
-			return nil, nil, errors.Errorf("no snapshot found to restore volume %s from", pvName)
-		}
-
-		restoreFromSnapshot = true
+	if sr.snapshotService == nil {
+		return nil, nil, errors.New("you must configure a persistentVolumeProvider to restore PersistentVolumes from snapshots")
 	}
-	if restore.Spec.RestorePVs == nil && sr.snapshotService != nil {
-		// when RestorePVs = Auto, don't error if the backup doesn't have snapshots
-		if backup.Status.VolumeBackups == nil || backup.Status.VolumeBackups[pvName] == nil {
-			return obj, nil, nil
-		}
 
-		restoreFromSnapshot = true
+	metadata, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to get a metadata accessor")
 	}
 
-	if restoreFromSnapshot {
-		backupInfo := backup.Status.VolumeBackups[pvName]
+	pvName := metadata.GetName()
 
-		volumeID, err := sr.snapshotService.CreateVolumeFromSnapshot(backupInfo.SnapshotID, backupInfo.Type, backupInfo.AvailabilityZone, backupInfo.Iops)
-		if err != nil {
-			return nil, nil, err
-		}
-
-		if err := kubeutil.SetVolumeID(spec, volumeID); err != nil {
-			return nil, nil, err
-		}
+	// If there's no snapshot for this PV, return an error
+	backupInfo, found := backup.Status.VolumeBackups[pvName]
+	if !found {
+		return nil, nil, errors.Errorf("no snapshot found to restore volume %s from", pvName)
 	}
 
-	var warning error
+	volumeID, err := sr.snapshotService.CreateVolumeFromSnapshot(backupInfo.SnapshotID, backupInfo.Type, backupInfo.AvailabilityZone, backupInfo.Iops)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	if sr.snapshotService == nil && len(backup.Status.VolumeBackups) > 0 {
-		warning = errors.New("unable to restore PV snapshots: Ark server is not configured with a PersistentVolumeProvider")
+	updatedObj, err := sr.snapshotService.SetVolumeID(obj, volumeID)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	return obj, warning, nil
+	return updatedObj, nil, nil
```

Comment:
Looks like we can get rid of the `warning` return arg since it's always nil

Reply:
I'll do that when I rebase